openstates/dc/bills.py

```diff
@@ -1,212 +1,449 @@
+            passed = True
+
+        vote = Vote("upper",date,"Committee Vote",passed,yes_count,no_count,other_count)
         bill.add_vote(vote)
 
-    def scrape(self, session, chambers):
-        url = 'http://dcclims1.dccouncil.us/lims/print/list.aspx?FullPage=True&Period=' + session
-
-        html = self.urlopen(url)
-        doc = lxml.html.fromstring(html)
-        rows = doc.xpath('//table/tr')
-        # skip first row
-        for row in rows[1:]:
-            bill_id, title, _ = row.xpath('td/text()')
-            title = title.replace(u'\xe2\x80\x99', "'")  # smart apostrophe
-
-            if bill_id.startswith('B'):
-                type = 'bill'
-            elif bill_id.startswith('CER'):
-                type = 'resolution'
-            elif bill_id.startswith('CA'):
-                type = 'contract'
-            elif bill_id.startswith('PR'):
-                type = 'resolution'
-            # don't collect these
-            elif (bill_id.startswith('GBM') or
-                  bill_id.startswith('HFA') or
-                  bill_id.startswith('IG')):
-                continue
+
+    def add_documents(self,attachment_path,bill,is_version=False):
+        global bill_versions
+        base_url = "http://lims.dccouncil.us/Download/" #nothing is actual links. we'll have to concatenate to get doc paths (documents are hiding in thrice-stringified json. eek.)
+        for a in attachment_path:
+            doc_type = a["Type"]
+            doc_name = a["Name"]
+            rel_path = a["RelativePath"]
+            if doc_type and doc_name and rel_path:  
+                doc_url = base_url+rel_path+"/"+doc_name
             else:
-                # will break if type isn't known
-                raise ValueError('unknown bill type: %s' % bill_id)
+                self.logger.warning("Bad link for document {}".format(doc_name))
+                return
+
+            mimetype = "application/pdf" if doc_name.endswith("pdf") else None
+
+            #figure out if it's a version from type/name
+            possible_version_types = ["SignedAct","Introduction","Enrollment","Engrossment"]
+            for vt in possible_version_types:
+                if vt.lower() in doc_name.lower():
+                    is_version = True
+                    doc_type = vt 
+
+            if "amendment" in doc_name.lower():
+                doc_type = "Amendment"
+
+            if is_version:
+                if doc_url in bill_versions:
+                    self.logger.warning("Version {} has been seen multiple times. Keeping first version seen".format(doc_url))
+                else:
+                    bill.add_version(doc_type,doc_url,mimetype=mimetype)
+                    bill_versions.append(doc_url)
+                continue
+                    
+
+            bill.add_document(doc_type,doc_url,mimetype=mimetype)
 
-            bill = Bill(session, 'upper', bill_id, title, type=type)
-            self.scrape_bill(bill)
```

Comment:
-            self.scrape_bill(bill)

-            bill = Bill(session, 'upper', bill_id, title, type=type)

 

+            bill.add_document(doc_type,doc_url,mimetype=mimetype)

+

+                    

+                continue

+                    bill_versions.append(doc_url)

+                    bill.add_version(doc_type,doc_url,mimetype=mimetype)

+                else:

+                    self.logger.warning("Version {} has been seen multiple times. Keeping first version seen".format(doc_url))

+                if doc_url in bill_versions:

+            if is_version:

+

+                doc_type = "Amendment"

+            if "amendment" in doc_name.lower():

+

+                    doc_type = vt 

+                    is_version = True

+                if vt.lower() in doc_name.lower():

+            for vt in possible_version_types:

+            possible_version_types = ["SignedAct","Introduction","Enrollment","Engrossment"]

+            #figure out if it's a version from type/name

+

+            mimetype = "application/pdf" if doc_name.endswith("pdf") else None

+

+                return

+                self.logger.warning("Bad link for document {}".format(doc_name))

-                raise ValueError('unknown bill type: %s' % bill_id)

-                # will break if type isn't known

             else:

+                doc_url = base_url+rel_path+"/"+doc_name

+            if doc_type and doc_name and rel_path:  

+            rel_path = a["RelativePath"]

+            doc_name = a["Name"]

+            doc_type = a["Type"]

+        for a in attachment_path:

+        base_url = "http://lims.dccouncil.us/Download/" #nothing is actual links. we'll have to concatenate to get doc paths (documents are hiding in thrice-stringified json. eek.)

+        global bill_versions

+    def add_documents(self,attachment_path,bill,is_version=False):

+

-                continue

-                  bill_id.startswith('IG')):

-                  bill_id.startswith('HFA') or

-            elif (bill_id.startswith('GBM') or

-            # don't collect these

-                type = 'resolution'

-            elif bill_id.startswith('PR'):

-                type = 'contract'

-            elif bill_id.startswith('CA'):

-                type = 'resolution'

-            elif bill_id.startswith('CER'):

-                type = 'bill'

-            if bill_id.startswith('B'):

-

-            title = title.replace(u'\xe2\x80\x99', "'")  # smart apostrophe

-            bill_id, title, _ = row.xpath('td/text()')

-        for row in rows[1:]:

-        # skip first row

-        rows = doc.xpath('//table/tr')

-        doc = lxml.html.fromstring(html)

-        html = self.urlopen(url)

-

-        url = 'http://dcclims1.dccouncil.us/lims/print/list.aspx?FullPage=True&Period=' + session

-    def scrape(self, session, chambers):

 

         bill.add_vote(vote)

+        vote = Vote("upper",date,"Committee Vote",passed,yes_count,no_count,other_count)

+

+            passed = True

+        if yes_count > no_count:

+        passed = False

+

+                other_count += vote_count

-                vote.other(member)

             else:

+                no_count = vote_count

+            elif v["VoteType"] == "No":

+                yes_count = vote_count

+            if v["VoteType"] == "Yes":

+

+            vote_count = 0 if v["VoteCount"] == "" else int(v["VoteCount"])

+        for v in vote_info:

-                vote.no(member)

-            elif 'NO' in vote_text:

-                vote.yes(member)

-            if 'YES' in vote_text:

-            vote_text = member_u.xpath('../../i/text()')[0].upper()

-            # normalize case

-            member = member_u.text

-        for member_u in doc.xpath('//u'):

-        # members are only text on page in a <u> tag

-

-        vote.add_source(url)

-

-                    no_count, other_count, voice_vote=voice)

-        vote = Vote('upper', vote_date, vote_type, passed, yes_count,

-

-                '//span[@id="VoteCount%s"]/b/text()' % n)[0])

-            other_count += extract_int(doc.xpath(

-        for n in xrange(3, 9):

         other_count = 0

 

+        date = self.date_format(date)

+            return

+            self.logger.warning("Committee vote has no data. Skipping.")

+        except KeyError:

+

+            vote_info = committee_action["Vote"]

+            date = committee_action["ActionDate"]

+        try:

+    def process_committee_vote(self,committee_action,bill):

-            '//span[@id="VoteCount2"]/b/text()')[0])

-        no_count = extract_int(doc.xpath(

-            '//span[@id="VoteCount1"]/b/text()')[0])

-        yes_count = extract_int(doc.xpath(

 

+        

-                    format(bill['bill_id']))

-                    "Not clear whether vote passed or failed: {}".

-            raise AssertionError(

-        if (not passed and not failed) or (passed and failed):

-        failed = (doc.xpath('//span[@id="VoteResultDisapproved"]//text()') == ['x'])

-        passed = (doc.xpath('//span[@id="VoteResultApproved"]//text()') == ['x'])

 

+        bill.add_vote(v)

-                    format(bill['bill_id']))

-                    "Not clear whether vote is voice or roll call: {}".

-            raise AssertionError(

-        if (not voice and not roll) or (voice and roll):

-        roll = (doc.xpath('//span[@id="VoteTypeRoll"]//text()') == ['x'])

-        voice = (doc.xpath('//span[@id="VoteTypeVoice"]//text()') == ['x'])

-        # check if voice vote / approved boxes have an 'x'

 

+            self.add_documents(vote["AttachmentPath"],bill,is_version)

-        vote_date = convert_date(doc.get_element_by_id('VoteDate').text)

 

+                is_version = True

+            if motion in ["enrollment","engrossment","introduction"]:

-        doc = lxml.html.fromstring(html)

-        html = self.urlopen(url)

 

+                pass

+            except KeyError:

+                    is_version = True

+                if vote["DocumentType"] in ["enrollment","engrossment","introduction"]:

+            try:

+            is_version = False

+        if "AttachmentPath" in vote:

+        #some documents/versions are hiding in votes.

+

+            vote["type"] = t.replace("bill:","")

+        elif "reading" in t:

+            vote["type"] = "amendment"

+        if "amendment" in t:

+

+                        type=t)

+                        date,

+                        motion,

+        bill.add_action("council",

+        

+            t = "other"

+        else:

+            t = "bill:reading:3"

+        elif "final reading" in motion.lower():

+            t = "bill:reading:3"

+        elif "3rd reading" in motion.lower():

+            t = "bill:reading:3"

+        elif "third reading" in motion.lower():

+            t = "bill:reading:2"

+        elif "2nd reading" in motion.lower():

+            t = "bill:reading:2"

+        elif "second reading" in motion.lower():

+            t = "bill:reading:1"

+        elif "1st reading" in motion.lower():

+            t = "bill:reading:1"

+        elif "first reading" in motion.lower():

+                t = "other" #we don't really have a thing for postponed bills

+        elif result in ["Tabled","Postponed"]:

+                t = "amendment:failed"

+            else:

+                t = "amendment:tabled"

+            elif result in ["Tabled","Postponed"]:

+                t = "amendment:passed"

+            if status:

+        if "amendment" in motion.lower():

-        url = base_url % (vote_type_id, bill['bill_id'])

-        base_url = 'http://dcclims1.dccouncil.us/lims/voting.aspx?VoteTypeID=%s&LegID=%s'

-    def scrape_vote(self, bill, vote_type_id, vote_type):

 

+        #and also add relevant actions

+        #level in the json, so we'll deal with them here

+        #the documents for the readings are inside the vote

 

+        

-        self.save_bill(bill)

-        bill['actions'] = sorted(bill['actions'], key=lambda b:b['date'])

 

+                v['other_votes'].append(mem_name)

+                v['other_count'] += 1

+            else:

+                v['no_votes'].append(mem_name)

+                v['no_count'] += 1

+            elif leg_vote["Vote"] == "2":

+                v['yes_votes'].append(mem_name)

+                v['yes_count'] += 1

+            if leg_vote["Vote"] == "1":

+            mem_name = member_ids[int(leg_vote["MemberId"])]

+        for leg_vote in leg_votes:

+                yes_votes=[],no_votes=[],other_votes=[])

+        v = Vote('upper',date,motion,status,0,0,0,

+        leg_votes = vote["MappedJASON"] #their typo, not mine

+

+        date = self.date_format(vote["DateOfVote"])

+

+                return

+                self.logger.warning("Unexpected vote result '{result},' skipping vote.".format(result=result))

+            except KeyError:

+                status = statuses[result.strip().lower()]

-                self.warning(str(e))

-            except scrapelib.HTTPError as e:

-                self.scrape_vote(bill, vote_type_id, vote_type)

             try:

+

+                    "passed":True}

+                    "declined":False,

+                    "failed":False,

+                    "disapproved":False,

+            statuses = {"approved":True,

+

+            result = vote["VoteResult"]    

+

+        else:

+                return

+                self.logger.warning("Could not find result of vote, skipping.")

+            else:

+                status = True

+                result = "Tabled"

+            elif "tabled" in motion.lower():

+                status = True #because we're talking abtout the motion, not the amendment

+                result = "Postponed"

+            if "postponed" in motion.lower():

+        if not "VoteResult" in vote:

+

+            return

+            self.logger.warning("Can't even figure out what we're voting on. Skipping.")

+        except KeyError:

+            motion = vote["ReadingDescription"]

+        try:

+    def process_vote(self, vote, bill, member_ids):

+

+

+        return member_dict

+

+                member_dict[session_id][member_id] = member_name

+                member_name = member["MemberName"]

+                member_id = int(member["ID"]) #should be ints already, but once one wasn't

+            for member in members:

+            member_dict[session_id] = {}

+        for session_id, members in member_data.items():

+        member_data = self.decode_json(response.json()['d'])["Members"]

+        response = self.post(search_data_url,headers={"Content-Type":"application/json"})

+        search_data_url = "http://lims.dccouncil.us/_layouts/15/uploader/AdminProxy.aspx/GetPublicSearchData"

+        member_dict = {} #three levels: from session to member_id to name

+    def get_member_ids(self):

+    

+            data = response["aaData"]

+            response = self.decode_json(response.json()["d"])

+            response = self.post(url,headers=headers,data=param_json)

+            param_json = json.dumps(params)

+            params["request"]["iDisplayStart"] = start_record

+            start_record += per_page

+            #get next page

+            

+                self.save_bill(bill)

+                bill.add_source(bill_source_url)

+

+                            self.add_documents(vote["AttachmentPath"],bill,is_version)

+                        if "AttachmentPath" in committee_info:

+                            self.process_committee_vote(committee_action,bill)

+                        for committee_action in committee_info:

+                    if len(committee_info) > 0:

+                    committee_info = bill_info["ComiteeMarkup"]

+                if "ComiteeMarkup" in bill_info: #their typo, not mine

+                #deal with committee votes

+

+     

+                    self.process_vote(vote, bill, member_ids)

+                for vote in votes:

+                votes = bill_info["VotingSummary"]

+                #full council votes

+

+

+                    self.add_documents(legislation_info["AttachmentPath"],bill)

+                if "AttachmentPath" in legislation_info:

+

+                    self.add_documents(legislation_info["MemoLink"],bill)

+                if "MemoLink" in legislation_info:

+

+                        self.logger.warning("Document path missing from 'Other Documents'")

+                    else:

+                        self.add_documents(d["AttachmentPath"],bill)

+                    if "AttachmentPath" in d:

+                for d in docs:

+                docs = bill_info["OtherDocuments"]

+                #deal with random docs floating around

+

+                                    type="other")

+                                    committees=committees,

+                                    date,

+                                    "comments from committee",

+                    bill.add_action("committee",

+                        committees.append(committee["Name"])

+                    for committee in legislation_info["CommitteeReferralComments"]:

+                    committees = []

+                if "CommitteeReferralComments" in legislation_info:

+

+                                    type="committee:referred")

+                                    committees=committees,

+                                    date,

+                                    "referred to committee",

+                        bill.add_action("committee",

+                    if committees != []:

+                            committees.append(committee["Name"])

-            # some votes randomly break

-                                     td.get('onclick')).groups()[0]

-            vote_type_id = re.search(r"LoadVotingInfo\(this\.id, '(\d+)'",

-            vote_type = td.text

-        for td in vote_tds:

-        vote_tds = doc.xpath('//td[starts-with(@id, "VoteTypeRepeater")]')

-        # votes

-

-                            bill.add_action(actor, action, date, type=atype)

                         else:

+                            break

+                            committees = []

+                        if committee["Name"].lower() == "retained by the council":

+                    for committee in legislation_info["ComitteeReferral"]:

+                    committees = []

+                if "ComitteeReferral" in legislation_info: #their typo, not mine

+                date = self.date_format(date)

+                    continue

+                    self.logger.warning("Crap, we can't find anything that looks like an action date. Skipping")

-                                         (action, date, bill['bill_id']))

-                            self.warning('could not convert %s %s [%s]' %

-                        if not isinstance(date, datetime.datetime):

-                        date = convert_date(date)

-                                 else 'upper')

-                        actor = ('mayor' if action.endswith('by Mayor')

-                                  'NOTCONSIDERED'):

-                    if date not in ('Not Signed', 'NOT CONSIDERED',

                 else:

+                    date = legislation_info["IntroductionDate"]

+                elif "IntroductionDate" in legislation_info:

+                    date = legislation_info["DateRead"]

+                if "DateRead" in legislation_info:

+                #deal with committee actions

+

+

+                                    transmitted_date)

+                                    "Transmitted to Congress for review",

+                        bill.add_action("US Congress",

+                        transmitted_date = self.date_format(congress["TransmittedDate"])

+                    if "TransmittedDate" in congress:

+                    congress = congress[0]

+                if len(congress) > 0:

+                congress = bill_info["CongressReview"]

+

+                        self.add_documents(mayor["AttachmentPath"],bill)

+                        #documents relating to the mayor's review

+                    if 'AttachmentPath' in mayor:

+

+                                        type="bill:veto_override:passed")

+                                        override_date,

+                                        "veto override",

+                            bill.add_action("upper",

+

+                            override_date = self.date_format(mayor["EnactedDate"])

+                        if 'EnactedDate' in mayor: #if it was returned and enacted but not signed, there was a veto override

+

+                                        type="governor:vetoed")

+                                        veto_date,

+                                        "vetoed",

+                        bill.add_action("mayor",

+

+                        veto_date = self.date_format(mayor["ReturnedDate"])

+                    elif 'ReturnedDate' in mayor: #if returned but not signed, it was vetoed

+

+

+                                        type="governor:signed")

+                                        signed_date,

+                                        "signed",

+                        bill.add_action("mayor",

+

+                        signed_date = self.date_format(mayor["SignedDate"])

+                    if 'SignedDate' in mayor:

+

+                                    type = "governor:received")

+                                    transmitted_date,

+                                    "transmitted to mayor",

+                        bill.add_action("mayor",

+

+                        transmitted_date = self.date_format(mayor["TransmittedDate"])

+                    if "TransmittedDate" in mayor:

+                    #in dc, mayor == governor because openstates schema

+

+                    mayor = mayor[0]

+                if mayor != []:

+                mayor = bill_info["MayorReview"]

+                #deal with actions involving the mayor

+

+

+                                    "bill:withdrawn")

+                                    withdrawn_date,

+                                    "withdrawn",

+                    bill.add_action(withdrawn_by,

+

+                    withdrawn_by = legislation_info["WithdrawnBy"][0]["Name"].strip()

+                    withdrawn_date = self.date_format(legislation_info["WithDrawnDate"])

+                if "WithDrawnDate" in legislation_info:

+

+                        bill["additional_information"] = add_info

+                    elif add_info:

+                        bill.add_title(prev_title.upper())

+                        prev_title = add_info.lower().replace("previously","").strip().replace(" ","")

+                    if "previously" in add_info.lower():

+                    add_info = legislation_info["AdditionalInformation"]

+                if "AdditionalInformation" in legislation_info:

+                #sometimes AdditionalInformation has a previous bill name

+

+                        bill.add_title(act_num)

+                    if act_num:

+                    act_num = legislation_info["ActNumber"]

+                if "ActNumber" in legislation_info:

+                #also sometimes it's got an act number

+

+                        bill.add_title(law_num)

+                    if law_num:

+                    law_num = legislation_info["LawNumber"]

+                if "LawNumber" in legislation_info:

+                #if it's become law, add the law number as an alternate title

+

+

+                    bill.add_sponsor(name=sponsor_name,type="cosponsor")

+                        sponsor_name = "Phil Mendelson"

+                    if sponsor_name == "Phil Pmendelson":

+                    sponsor_name = s["Name"]

+                for s in cosponsors:

+                    bill.add_sponsor(name=sponsor_name,type="primary")

+                        sponsor_name = "Phil Mendelson"

+                    if sponsor_name == "Phil Pmendelson":

+                    #they messed up Phil Mendelson's name

+                    sponsor_name = i["Name"]

+                for i in introducers:

+                    cosponsors = []

+                except KeyError:

+                    cosponsors = legislation_info["CoSponsor"]

+                    #sometimes there are cosponsors, sometimes not.

+                try:

+                introducers = legislation_info["Introducer"]

+                #sponsors and cosponsors

+

+                bill = Bill(session,"upper", bill_id, title, type=bill_type)

+                #dc has no chambers. calling it all upper

+                

+                    bill_type = "bill"

+                else:

+                    bill_type = "resolution"

+                if bill_id.startswith("R") or bill_id.startswith("CER"):

+                

+                

+                

+                title = legislation_info["ShortTitle"]

+                legislation_info = bill_info["Legislation"][0]

+

+

+                bill_source_url = "http://lims.dccouncil.us/Legislation/"+bill_id

+                bill_info = self.decode_json(bill_info.json()["d"])["data"]

+                bill_info = self.post(bill_url,headers=headers,data=json.dumps(bill_params))

+                bill_params = {"legislationId":bill_id}

+                    continue

+                    #actually an agenda, skip

+                if bill_id.startswith("AG"):

+                bill_id = bill["Title"]

+

+                bill_versions = [] #sometimes they're in there more than once, so we'll keep track

+            for bill in data:

+

+        while len(data) > 0:

+

+        global bill_versions

+        

+        data = response["aaData"]

+        response = self.decode_json(response.json()["d"])

+        #the response is a terrible string-of-nested-json-strings. Yuck.

+        response = self.post(url,headers=headers,data=param_json)

+        param_json = json.dumps(params)

+        params = {"request":{"sEcho":2,"iColumns":4,"sColumns":"","iDisplayStart":0,"iDisplayLength":per_page,"mDataProp_0":"ShortTitle","mDataProp_1":"Title","mDataProp_2":"LegislationCategories","mDataProp_3":"Modified","iSortCol_0":0,"sSortDir_0":"asc","iSortingCols":0,"bSortable_0":"true","bSortable_1":"true","bSortable_2":"true","bSortable_3":"true"},"criteria":{"Keyword":"","Category":"","SubCategoryId":"","RequestOf":"","CouncilPeriod":str(session),"Introducer":"","CoSponsor":"","ComitteeReferral":"","CommitteeReferralComments":"","StartDate":"","EndDate":"","QueryLimit":100,"FilterType":"","Phases":"","LegislationStatus":"0","IncludeDocumentSearch":"false"}}

+        bill_url = "http://lims.dccouncil.us/_layouts/15/uploader/AdminProxy.aspx/GetPublicData"

+        url = "http://lims.dccouncil.us/_layouts/15/uploader/AdminProxy.aspx/GetPublicAdvancedSearch"

+        headers = {"Content-Type":"application/json"}

+

+        start_record = 0

+        per_page = 10 #seems like it gives me 10 no matter what.

+        member_ids = self.get_member_ids()[session]

+        #get member id matching for vote parsing

+    def scrape(self, session, chambers):

+

+

-                # actions that mean nothing happened

-

-                        break

-                                            type=subatype)

-                            bill.add_action('upper', sub_action, date,

-                        if date:

-                        date = convert_date(date)

-                    if date.startswith(prefix):

-                for prefix, sub_action, subatype in subactions:

-                # check if the action starts with a subaction prefix

-

-            if date:

-            date = doc.get_element_by_id(elem_id).text

-        for action, elem_id, atype in actions:

-

-        )

-            ('DEEMED DISAPPROVED', 'Deemed disapproved without council action', 'bill:failed'),

-            ('DEEMED APPROVED', 'Deemed approved without council action', 'bill:passed'),

-            ('TABLED', 'Tabled', 'other'),

-            ('WITHDRAWN BY', 'Withdrawn', 'bill:withdrawn'),

-        subactions = (

-

-        )

-            ('Re-transmitted to Congress', 'DateReTransmitted', 'other'),

-            ('Transmitted to Congress', 'DateTransmittedCongress', 'other'),

-            ('Vetoed by Mayor', 'DateVeto', 'governor:vetoed'),

-            ('Enacted', 'DateEnactment', 'other'),

-            ('Veto Override', 'DateOverride', 'bill:veto_override:passed'),

-            ('Returned by Mayor', 'DateReturned', 'other'),

-            ('Signed by Mayor', 'DateSigned', 'governor:signed'),

-            ('Transmitted to Mayor', 'DateTransmittedMayor', 'governor:received'),

-            ('Reconsideration', 'DateReconsideration', 'other'),

-            ('Third Vote', 'DateThirdVote', ['bill:reading:3']),

-            ('Final Vote', 'DateFinalVote', 'bill:reading:3'),

-            ('First Vote', 'DateFirstVote', 'bill:reading:1'),

-            ('Committee Action', 'DateCommitteeAction', 'other'),

-            ('Introduction', 'DateIntroduction', 'bill:introduced'),

-        actions = (

-        # actions

-

-                bill.add_sponsor('cosponsor', cosponsor.strip())

-            for cosponsor in cosponsored_by.split(','):

-        if cosponsored_by:

-        cosponsored_by = doc.get_element_by_id('CoSponsoredBy').text

-

-                             official_type='requestor')

-            bill.add_sponsor(stype, requested_by.strip(),

-            stype = 'primary' if not introduced_by else 'cosponsor'

-        if requested_by:

-        requested_by = doc.get_element_by_id('RequestedBy').text

-

-                bill.add_sponsor('primary', sponsor.strip())

-            for sponsor in introduced_by.split(', '):

-        if introduced_by:

-        introduced_by = doc.get_element_by_id('IntroducedBy').text

-        # sponsors

-

-                self.warning('duplicate version on %s' % bill_url)

-            except ValueError as e:

-                                 mimetype='application/pdf')

-                bill.add_version(link.text, link.get('href'),

-            try:

-        for link in doc.xpath('//a[starts-with(@id, "DocumentRepeater")]'):

-        # get versions

 

+        return stringy_json

+                return self.decode_json(json.loads(stringy_json))

+            if len(stringy_json) > 0 and stringy_json[0] in ["[","{",u"[",u"{"]:

+        elif type(stringy_json) in (str,unicode):

-        doc.make_links_absolute(bill_url)

-        doc = lxml.html.fromstring(bill_html)

-        bill_html = self.urlopen(bill_url)

 

+                stringy_json[i] = self.decode_json(stringy_json[i])

+            for i in range(len(stringy_json)):

+        elif type(stringy_json) == list:

-        bill.add_source(bill_url)

 

+                stringy_json[key] = self.decode_json(stringy_json[key])

+            for key in stringy_json:

+        if type(stringy_json) == dict:

+        #the "json" they send is recursively string-encoded.

+    def decode_json(self,stringy_json):

-        bill_url = 'http://dcclims1.dccouncil.us/lims/legislation.aspx?LegNo=' + bill['bill_id']

-    def scrape_bill(self, bill):

 

     jurisdiction = 'dc'

 class DCBillScraper(BillScraper):

 

-

-        return None

-    except ValueError:

-        return datetime.datetime.strptime(text, '%A, %B %d, %Y')

-

-            return datetime.datetime.strptime(short_date[0], '%m-%d-%y')

-        if short_date:

-        short_date = re.findall('\d{1,2}-\d{1,2}-\d{2}', text)

-            return datetime.datetime.strptime(long_date[0], '%m-%d-%Y')

-        if long_date:

-    try:

-    long_date = re.findall('\d{1,2}-\d{1,2}-\d{4}', text)

-def convert_date(text):

-

-    return int(text.replace(u'\xc2', '').strip())

-def extract_int(text):

 

 from billy.scrape.votes import Vote

 from billy.scrape.bills import BillScraper, Bill

 

 import scrapelib

 

+import json

 import lxml.html

 import datetime

 import re

@@ -1,212 +1,449 @@